### ISSUE TITLE
sql: localities can't be changed seamlessly if there are replica constraints

### ISSUE THREAD

**Issue by jordanlewis** [cockroachdb/cockroach#42715]
A user had the following locality hierarchy:

`continent, region`

The user had leaseholder preferences bound to regions, and replica constraints bound to continents.

The user wished to migrate the continent key to a different name, and decided to do this by restarting all nodes with another locality pair (we'll call it `continent2`). So, the new hierarchy was:

`continent2, continent, region`

When the user restarted, the cluster experienced leaseholder thrashing. The current hypothesis is that this is because the replica constraints don't apply since `continent` is no longer the top level locality.

Whatever the reason, this is not the best experience for a user. We should make it possible to do more locality changes online here.

**Comment by nvanbenschoten** [comment]
Out of curiosity, would restarting with `continent, continent2, region` have worked? That doesn't solve the general case, but I wonder if it's a workaround here.

**Comment by roncrdb** [comment]
@nvanbenschoten 

They tried a few different versions: 

```
Unstable leaseholders & replicas:
continent2=us,continent=us,region=us-west
continent=us, continent2 =us,region=us-west
continent=us,foo=bar,region=us-west
continent=us,region=us-west,foo=bar
```

**Comment by nvanbenschoten** [comment]
What does "unstable" mean? Is it possible that any worked and the system was just correcting itself again?

**Comment by roncrdb** [comment]
@nvanbenschoten I asked the user directly, the thrashing they saw never corrected itself until they intervened and reverted the change they made to the locality key. 

What do we need from them, or what are some next steps to take here?

**Comment by nvanbenschoten** [comment]
@roncrdb having read the interaction, I'm not convinced it confirms that none of these approaches work. There doesn't seem to be a strong indication (but would love to be proved wrong) that all variations were given sufficient time to converge.

I don't think we need anything more from them though. We can test this ourselves.

**Comment by roncrdb** [comment]
@nvanbenschoten 

Sounds good!

**Comment by RoachietheSupportRoach** [comment]
Zendesk ticket #[4133](https://cockroachdb.zendesk.com/agent/tickets/4133) has been linked to this issue.

**Comment by nvanbenschoten** [comment]
I'm not able to reproduce this, even with the locality set to `continent2, continent, region`. Here's the script I was using to test:
```bash
# Create cluster.
export CLUSTER=local
roachprod create ${CLUSTER} --nodes=10
roachprod stage  ${CLUSTER} cockroach

# Start cluster.
roachprod start ${CLUSTER}:1-3 --args='--locality=continent=us,region=us-west'
roachprod start ${CLUSTER}:4-6 --args='--locality=continent=us,region=us-east'
roachprod start ${CLUSTER}:7-9 --args='--locality=continent=eu,region=eu-west'

# Set up ranges with zone configs
roachprod run ${CLUSTER}:10 -- './cockroach workload init kv --splits=1000 {pgurl:1}'
roachprod run ${CLUSTER}:1  -- ./cockroach sql --insecure -e "ALTER DATABASE kv CONFIGURE ZONE USING constraints = '[+continent=us]', lease_preferences = '[[+region=us-west]]'"
roachprod run ${CLUSTER}:1  -- ./cockroach sql --insecure -e "SELECT range_id, replicas, lease_holder FROM crdb_internal.ranges ORDER BY 1;"

# Wait for the cluster to stabilize.

# Restart cluster with new locality.
# Options tried => effect:
## continent2=us,continent=us,region=us-west => no replica movement or lease instability
## continent=us,continent2=us,region=us-west => no replica movement or lease instability
## continent=not-us,region=us-west           => replica movement and lease instability
for i in {1..3}; do
    roachprod stop ${CLUSTER}:${i}; 
    roachprod start ${CLUSTER}:${i} --args='--locality=continent=us,continent2=us,region=us-west'; 
    sleep 10; 
done
```

This lines up with the code, which indicates that the ordering of localities shouldn't matter:
https://github.com/cockroachdb/cockroach/blob/45d709f1bb65c26573a7b42de13df15510c1a5ce/pkg/config/zone.go#L469-L473

So I'm not sure what's going on here. I tested with 19.2 so maybe there's a difference between that and 19.1. Or maybe there's something going on when load is continually applied to the cluster?

The other option is that there's something different between required constraints (`[+continent=us]`) and per-replica constraints (`{+continent=us: 2}`) that's going wrong here. I don't know what that could be from the code.

**Comment by nvanbenschoten** [comment]
The other thing I'm thinking that could have gone wrong is that lease preferences alone don't guarantee that a replica is in the preferred region. In other words, `constraints = '{+continent=us: 2}', lease_preferences = '[[+region=us-west]]'` doesn't guarantee that there will actually be a replica in `us-west` to satisfy the lease preference. For that, you'd need something like `constraints = '{+continent=us: 2, +region=us-west: 1}', lease_preferences = '[[+region=us-west]]'`. Maybe that provided some opportunity for leases to destabilize during the rolling upgrade?

**Comment by nvanbenschoten** [comment]
> The other option is that there's something different between required constraints ([+continent=us]) and per-replica constraints ({+continent=us: 2}) that's going wrong here.

Nope, this seems to work fine.

**Comment by nvanbenschoten** [comment]
I gave this another shot with full cluster rolling upgrades between trials. When using localities like `continent=us,continent2=us,region=us-west`, I began to see replica movement that did not converge even as the number of replicas on each node stayed relatively stable. I also saw some amount of leaseholder movement that did not converge. These both line up with the full report that initially inspired this issue.

I tracked this down to rebalancing driven by range count limits being exceeded (reason `range-overfull`). This was being computed in [`shouldRebalance`](https://github.com/cockroachdb/cockroach/blob/a7cf65526700caca693fd059fda91646a2913190/pkg/storage/allocator_scorer.go#L827). Instrumenting the code while this was going on revealed that `sl.candidateRanges.mean` was set to 0, which guaranteed that we'd attempt a rebalance. Tracking this up through `rebalanceCandidates`, it eventually became clear that the `StoreList` assigned to the corresponding `comparableStoreList` was empty, which shouldn't happen. This shouldn't be possible because, at a minimum, the Store that currently owns the replica should always be considered a candidate to retain its replica. So why wasn't this being respected?

It turns out that `rangeDiversityScore` and `diversityRebalanceFromScore` were computing slightly different results for the diversity score of a replica before and after moving to a Store that it already lived on. This again shouldn't be possible and it wasn't immediately clear why it was. Looking closely, I saw diversity scores like `0.7777777777777777` vs. `0.7777777777777778`. This hinted at a floating point comparison bug. Sure enough, floating-point addition is not associative (under all inputs) and due to randomized map iteration order in Go, `rangeDiversityScore` and `diversityRebalanceFromScore` we computing slightly different results.

So because they were computing different results, we could get cases where `len(comparableCands) == 0` and this would, in turn, lead to `shouldRebalance` returning true by default. That's a weird interaction and if there's a precondition that `len(comparableCands) > 0` for the rest of `rebalanceCandidates` to work correctly, we should make the requirement explicit and assert that as soon as possible (cc. @darinpp). We should also assert that it's not possible to fall out of [this loop](https://github.com/cockroachdb/cockroach/blob/a7cf65526700caca693fd059fda91646a2913190/pkg/storage/allocator_scorer.go#L651) without having set a matching `StoreList`. I never saw that happen in practice, but it looks suspicious to me and is just asking for issues.

So that explains the bug, but why did adding the `continent2` locality tier trigger it? The answer to this goes back to floating-point arithmetic and under which circumstances it is associative. Floating-point arithmetic is only guaranteed to be associative if all intermediate results are representable without loss of precision. With that in mind, it turns out that `Locality.DiversityScore` was using the number of locality tiers in the denominator of a fractional term. So with 2 locality tiers, all intermediate results were still representable as 64-bit floating points. However, with 3 locality tiers, some intermediate results lost precision (e.g. 1/3) so the arithmetic performed on top of it in `rangeDiversityScore` and `diversityRebalanceFromScore` was no longer deterministic, given the randomized arithmetic order due to Go's randomize map iteration order. Here's an example showing the kind of thing we can see before and after adding the third locality tier: https://play.golang.org/p/FN57cbZKIv6.

So this is a real bug. @darinpp is looking into something very related in a different issue, so he's going to write a fix for this as well. Hopefully, the fix will be small enough to backport to at least v19.2.
